tests/bluetooth/tester/src/btp_vcp.c

```diff
@@ -224,197 +218,160 @@ static struct bt_aics_cb aics_cb = {
-void aics_mute_disable(void)
-{
-	LOG_DBG("AICS mute disable");
-
-	tester_rsp(BTP_SERVICE_ID_AICS, AICS_MUTE_DISABLE,
-		   CONTROLLER_INDEX, BTP_STATUS_FAILED);
+	return BTP_STATUS_SUCCESS;
 }
 
-void aics_desc(uint8_t *data)
+static uint8_t aics_desc(const void *cmd, uint16_t cmd_len,
+			 void *rsp, uint16_t *rsp_len)
 {
-	const struct aics_audio_desc *cmd = (void *) data;
+	const struct btp_aics_audio_desc_cmd *cp = cmd;
 	char description[BT_AICS_MAX_INPUT_DESCRIPTION_SIZE];
 
 	LOG_DBG("AICS description");
 
-	if (cmd->desc_len > BT_AICS_MAX_INPUT_DESCRIPTION_SIZE) {
-		LOG_ERR("Too long input (%u chars required)",
-			BT_AICS_MAX_INPUT_DESCRIPTION_SIZE);
-		goto rsp;
+	if (cmd_len < sizeof(*cp) ||
+	    cmd_len != sizeof(*cp) + cp->desc_len) {
+		return BTP_STATUS_FAILED;
+	}
+
+	if (cp->desc_len >= sizeof(description)) {
+		return BTP_STATUS_FAILED;
 	}
 
-	memcpy(description, cmd->desc, cmd->desc_len);
-	description[cmd->desc_len] = '\0';
+	memcpy(description, cp->desc, cp->desc_len);
+	description[cp->desc_len] = '\0';
 
 	for (int i = 0; i < CONFIG_BT_VCP_VOL_REND_AICS_INSTANCE_COUNT; i++) {
 		if (bt_aics_description_set(included.aics[i], description) != 0) {
-			tester_rsp(BTP_SERVICE_ID_AICS, AICS_DESCRIPTION,
-				   CONTROLLER_INDEX, BTP_STATUS_FAILED);
-			return;
+			return BTP_STATUS_FAILED;
 		}
 	}
-rsp:
-	tester_rsp(BTP_SERVICE_ID_AICS, AICS_DESCRIPTION, CONTROLLER_INDEX,
-		   BTP_STATUS_FAILED);
-}
-
-void tester_handle_aics(uint8_t opcode, uint8_t index, uint8_t *data, uint16_t len)
-{
-	switch (opcode) {
-	case AICS_READ_SUPPORTED_COMMANDS:
-		aics_supported_commands(data, len);
-		break;
-	case AICS_SET_GAIN:
-		aics_set_gain(data);
-		break;
-	case AICS_MUTE:
-		aics_mute();
-		break;
-	case AICS_UNMUTE:
-		aics_unmute();
-		break;
-	case AICS_MAN_GAIN:
-		aics_man_gain();
-		break;
-	case AICS_AUTO_GAIN:
-		aics_auto_gain();
-		break;
-	case AICS_MAN_GAIN_ONLY:
-		aics_auto_gain_only();
-		break;
-	case AICS_AUTO_GAIN_ONLY:
-		aics_auto_gain_only();
-		break;
-	case AICS_DESCRIPTION:
-		aics_desc(data);
-		break;
-	case AICS_MUTE_DISABLE:
-		aics_mute_disable();
-		break;
-	default:
-		tester_rsp(BTP_SERVICE_ID_AICS, opcode, index,
-			   BTP_STATUS_UNKNOWN_CMD);
-		break;
-	}
+
+	return BTP_STATUS_SUCCESS;
 }
 
+static const struct btp_handler aics_handlers[] = {
+	{
+		.opcode = BTP_AICS_READ_SUPPORTED_COMMANDS,
+		.index = BTP_INDEX_NONE,
+		.expect_len = 0,
+		.func = aics_supported_commands,
+	},
+	{
+		.opcode = BTP_AICS_SET_GAIN,
+		.expect_len = sizeof(struct btp_aics_set_gain_cmd),
+		.func = aics_set_gain,
+	},
+	{
+		.opcode = BTP_AICS_MUTE,
+		.expect_len = 0,
+		.func = aics_mute,
+	},
+	{
+		.opcode = BTP_AICS_UNMUTE,
+		.expect_len = 0,
+		.func = aics_unmute,
+	},
+	{
+		.opcode = BTP_AICS_MAN_GAIN,
+		.expect_len = 0,
+		.func = aics_man_gain,
+	},
+	{
+		.opcode = BTP_AICS_AUTO_GAIN,
+		.expect_len = 0,
+		.func = aics_auto_gain,
+	},
+	{
+		.opcode = BTP_AICS_DESCRIPTION,
+		.expect_len = -1,
```

Comment:
Can you explain the `-1` in length?

Reply:
> yeap, I'll add comments in structure definition. -1 means that length is variable and is validated by handler itself

Suggest to use a `#define` for that then :) A named value would make it much more clear